share/shwap/p2p/bitswap/block_store.go

```diff
@@ -0,0 +1,113 @@
+package bitswap
+
+import (
+	"context"
+	"fmt"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+
+	eds "github.com/celestiaorg/celestia-node/share/new_eds"
+	bitswappb "github.com/celestiaorg/celestia-node/share/shwap/p2p/bitswap/pb"
+)
+
+// Accessors abstracts storage system that indexes and manages multiple eds.Accessors by network height.
+type Accessors interface {
+	// Get returns an Accessor by its height.
+	Get(ctx context.Context, height uint64) (eds.Accessor, error)
+}
+
+// Blockstore implements generalized Bitswap compatible storage over Shwap containers
+// that operates with Block and accesses data through Accessors.
+type Blockstore struct {
+	Accessors
+}
+
+func (b *Blockstore) getBlock(ctx context.Context, cid cid.Cid) (blocks.Block, error) {
+	spec, ok := specRegistry[cid.Prefix().MhType]
+	if !ok {
+		return nil, fmt.Errorf("unsupported Block type: %v", cid.Prefix().MhType)
+	}
+
+	blk, err := spec.builder(cid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build a Block for %s: %w", spec.String(), err)
+	}
+
+	eds, err := b.Accessors.Get(ctx, blk.Height())
+	if err != nil {
+		return nil, fmt.Errorf("getting EDS Accessor for height %v: %w", blk.Height(), err)
+	}
+
+	if err = blk.Populate(ctx, eds); err != nil {
+		return nil, fmt.Errorf("failed to populate Shwap Block on height %v for %s: %w", blk.Height(), spec.String(), err)
+	}
+
+	containerData, err := blk.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("marshaling Shwap container: %w", err)
+	}
+
+	blkProto := bitswappb.Block{
+		Cid:       cid.Bytes(),
+		Container: containerData,
+	}
+
+	blkData, err := blkProto.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("marshaling Bitswap Block protobuf: %w", err)
+	}
+
+	bitswapBlk, err := blocks.NewBlockWithCid(blkData, cid)
```

Comment:
It seems to marshal cid in proto and then add it second time in NeBlockWithCid. Does it contain it twice?

Reply:
Sort of. The first one is part of the data and is transferred over the wire to be unpacked in umarshal. The other one is needed for bitswap to identify the block, but is not actually transmitted over the wire

Reply:
Extracted proto wrapping funcs in separate files